## Analysis of DNA flexibility using neural network models
Here are listed all the python scripts and code used throughout my research project. Many of the scripts utilise much of the code produced in the Cyclizablility prediction code script, with partial alterations to fit certain data or perform seperate analyses. 

__'Model Confirmation':__
Validating the neural network using a dataset of random sequences with known cyclizability 

__'Cyclizability Prediction Code':__
Original Code for predicting cyclizability around the Nucleosomes of schizosaccharomyces pombe and performing codon randomization 

__'Cerevisiae 1-9':__
Code used to predict cyclizability around cerevisiae nucleosomes and codon randomisation

__'Cyclizability 2000 bp of +1 Nucleosome':__
Mapping the cyclizability of the +1 nucleosome at a greater range 

__'Multiple Codon Randomisations':__ 
Performing multiple rounds of codon ranomisation on the +5 Nucleosome 

__'Finding Common Amino Acids Pombe':__
Exploring how the prevelance of amino acids change along a Nucleosome in pombe

__'Finding Common Amino Acids Cerevisiae':__
Exploring how the prevelance of amino acids change along a Nucleosome in cerevisiae

__'Analyses of Amino Acid cyclizability':__
Finding metrics of average amino acid cyclizability using codon averages and simulated peptides 

__'Finding Common Codons':__
Finding which codon's are most frequent in the genes of cerevisiae and pombe for use as weightings in amino acid analyses

__'Plotting Stuff':__
Loading data produced from other script and producing comparison plots and performing statistical comparions using T-tests or calculating SSD
